acq4/devices/PressureControl/device.py

```diff
@@ -31,17 +34,54 @@ def __init__(self, manager, config, name):
         self.sources = ("regulator", "user", "atmosphere")
 
     @Future.wrap
-    def attainPressure(self, source, maximum=None, minimum=None, _future=None):
-        self.setPressure(source)
-        while True:
-            pressure = self.getPressure()
-            if minimum is not None and pressure < minimum:
-                self.setPressure(source, pressure=minimum)
-            elif maximum is not None and pressure > maximum:
-                self.setPressure(source, pressure=maximum)
+    def attainPressure(
+        self,
+        source: str = "regulator",
+        target: Optional[float] = None,
+        target_tolerance: float = 10,
+        maximum: Optional[float] = None,
+        minimum: Optional[float] = None,
+        rate: Optional[float] = None,
+        _future: Optional[Future] = None,
+    ) -> None:
+        if target is None and maximum is None and minimum is None:
+            raise ValueError("Must specify at least one of target, maximum, or minimum")
+        if target is not None and (maximum is not None or minimum is not None):
+            raise ValueError("Cannot specify both target and maximum/minimum")
+        # TODO do we need to guarantee that the source gets set?
+
+        def value_is_out_of_bounds(val):
```

Comment:
Something like this? It's not dramatically shorter, but feels more straightforward to me.

```
if target is not None:
    minimum = target - target_tolerance
    maximum = target + target_tolerance

startPressure = endPressure = self.getPressure()
if minimum is not None:
    endPressure = max(minimum, endPressure)
if maximum is not None:
    endPressure = min(maximum, endPressure)

startTime = ptime.time()   # time.time() works differently on windows! 
endTime = (endPressure - startPressure) / rate
fracDone = 0
while fracDone < 1:
    fracDone = min((ptime.time() - startTime) / (endTime - startTime), 1)
    self.setPressure(startPressure + fracDone * (endPressure - startPressure))
    _future.sleep(self.regulatorSettlingTime) 
```